File: leetcode-js/212-word-search-ii.js
const A = 'a'.charCodeAt();

/**
 * @param  {character[][]} board
 * @param  {string[]}      words
 * @return {string[]}
 */
var findWords = function(board, words) {
    const res = [];
    const root = buildTrie(words);
    for (let i = 0; i < board.length; i += 1) {
        for (let j = 0; j < board[0].length; j += 1) {
            dfs(board, i, j, root, res);
        }
    }
    return res;
};

function dfs(board, i, j, p, res) {
    const c = board[i][j];
    if (c === '#' || !p.next[c.charCodeAt() - A]) {
        return;
    }
    p = p.next[c.charCodeAt() - A];
    if (p.word) {
        res.push(p.word);
        p.word = null;
    }
    board[i][j] = '#';
    if (i > 0) {
        dfs(board, i - 1, j, p, res);
    }
    if (j > 0) {
        dfs(board, i, j - 1, p, res);
    }
    if (i < board.length - 1) {
        dfs(board, i + 1, j, p, res);
    }
    if (j < board[0].length - 1) {
        dfs(board, i, j + 1, p, res);
    }
    board[i][j] = c;
}

function buildTrie(words) {
    const root = new TrieNode();
    for (const w of words) {
        let p = root;
        for (const c of w) {
            const i = c.charCodeAt() - A;
            if (!p.next[i]) {
                p.next[i] = new TrieNode();
            }
            p = p.next[i];
        }
        p.word = w;
    }
    return root;
}

class TrieNode {
    constructor() {
        this.next = [];
        this.word = null;
    }
}

console.log(findWords(["ab","cd"], ["ab","cb","ad","bd","ac","ca","da","bc","db","adcb","dabc","abb","acb"]))

File: leetcode-java/212-word-search-ii.java
public class Solution {
    class TrieNode {
        String word;
        TrieNode[] children = new TrieNode[26];
    }

    public List<String> findWords(char[][] board, String[] words) {
        TrieNode root = buildTrie(words);
        List<String> res = new LinkedList<>();
        int m = board.length;
        int n = board[0].length;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char c = board[i][j];
                if (root.children[c - 'a'] != null) {
                    dfs(board, m, n, root.children[c - 'a'], i, j, res);
                }
            }
        }

        return res;
    }

    TrieNode buildTrie(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                if (node.children[c - 'a'] == null) {
                    node.children[c - 'a'] = new TrieNode();
                }
                node = node.children[c - 'a'];
            }
            node.word = word;
        }
        return root;
    }

    void dfs(char[][] board, int m, int n, TrieNode trie, int i, int j, List<String> res) {
        if (trie.word != null) {
            res.add(trie.word);
            trie.word = null;
        }

        char original = board[i][j];
        board[i][j] = '#';

        if (i > 0 && board[i - 1][j] != '#' && trie.children[board[i - 1][j] - 'a'] != null) {
            dfs(board, m, n, trie.children[board[i - 1][j] - 'a'], i - 1, j, res);
        }
        if (i < m - 1 && board[i + 1][j] != '#' && trie.children[board[i + 1][j] - 'a'] != null) {
            dfs(board, m, n, trie.children[board[i + 1][j] - 'a'], i + 1, j, res);
        }
        if (j > 0 && board[i][j - 1] != '#' && trie.children[board[i][j - 1] - 'a'] != null) {
            dfs(board, m, n, trie.children[board[i][j - 1] - 'a'], i, j - 1, res);
        }
        if (j < n - 1 && board[i][j + 1] != '#' && trie.children[board[i][j + 1] - 'a'] != null) {
            dfs(board, m, n, trie.children[board[i][j + 1] - 'a'], i, j + 1, res);
        }

        board[i][j] = original;
    }
}
